### PR TITLE
fix(types): pass types through memoize properly

### PR DESCRIPTION
**Pull request by LeoMcA**
### Problem

While reviewing https://github.com/mdn/yari/pull/10433 it annoyed me that `memoize` wasn't passing types through properly, and we were losing types in most of `build/curriculum.ts` because of this.

### Solution

Pass through the types of the wrapped function in `memoize`, fix the things this breaks. (I suggest reviewing with ignore whitespace because one function gets majorly indented from these fixes).

---

## How did you test this change?

`yarn check:tsc`

### CODE REVIEW

File: kumascript/index.ts [hunk 1]
```diff
@@ -37,7 +37,7 @@ export async function render(
     selective_mode = false,
     invalidateCache = false,
   }: RenderOptions = {},
-  doc?: Doc
+  blogPost?: BlogPostDoc
```
Comment (caugner) [inline kumascript/index.ts:40]: This doesn't look right. Why would `kumascript.render()` return a `BlogPostDoc`?

File: kumascript/index.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ import {
   LIVE_SAMPLES_BASE_URL,
 } from "../libs/env/index.js";
 import { SourceCodeError } from "./src/errors.js";
-import { Doc } from "../libs/types/document.js";
+import { BlogPostDoc } from "../build/blog.js";
```
Comment (caugner) [inline kumascript/index.ts:17]: Btw should `BlogPostDoc` be moved into `libs/types`?

File: content/document.ts [hunk 3]
```diff
@@ -34,7 +34,7 @@ import {
   MEMOIZE_INVALIDATE,
 } from "./utils.js";
 import * as Redirect from "./redirect.js";
-import { DocFrontmatter } from "../libs/types/document.js";
+import { DocFrontmatter, UnbuiltDocument } from "../libs/types/document.js";
```
Comment (caugner) [inline content/document.ts:37]: Maybe also worth moving `BuiltDocument` into `libs/types` if it isn't already, and extracting common parts if any?

**Review by caugner — COMMENTED** [review]
Very nice, just skimmed over, and left some comments.

File: kumascript/index.ts (hunk 1 above)
Comment (LeoMcA) [reply kumascript/index.ts:40]: It doesn't, this is the last argument, the return argument is the line below (the Promise)

Comment (fiji-flo) [reply kumascript/index.ts:40]: That should stay called `doc` it just replaces the `Document.findByURL`. And the type is tricky. Probably `Doc` as `BlogPostDoc` should just extend `Doc` but that is utterly broken right now.

File: content/translations.ts [hunk 4]
```diff
@@ -8,7 +8,7 @@ const LANGUAGES = new Map(
   })
 );
 
-type Translation = {
+export type Translation = {
```
Comment (yin1999) [inline content/translations.ts:11]: We can replace the type declaration here, as there was a duplicated one in [`libs/types/document.ts`](https://github.com/mdn/yari/blob/006eb85729f34721eea92e5f2146f83dd5c37438/libs/types/document.ts#L111-L115)

File: kumascript/index.ts (hunk 1 above)
Comment (caugner) [reply kumascript/index.ts:40]: I don't think `Doc` applies here.

BlogPostDoc:

https://github.com/mdn/yari/blob/006eb85729f34721eea92e5f2146f83dd5c37438/build/blog.ts#L347-L355

DocMetadata/Doc:

https://github.com/mdn/yari/blob/006eb85729f34721eea92e5f2146f83dd5c37438/libs/types/document.ts#L128-L159

File: content/translations.ts (hunk 4 above)
Comment (caugner) [reply content/translations.ts]: Good catch, fixed in c00822d4e0.

File: content/document.ts (hunk 3 above)
Comment (caugner) [reply content/document.ts:37]: Never mind for now.

**Review by caugner — APPROVED** [review]
@LeoMcA Please have a look at the changes I pushed, but LGTM.

**Pull request merged**